codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/RestJsonProtocolGenerator.java

```
@@ -85,27 +85,33 @@ public void generateSharedComponents(GenerationContext context) {
     }
 
     @Override
-    protected void writeDefaultHeaders(GenerationContext context, OperationShape operation) {
-        super.writeDefaultHeaders(context, operation);
-        AwsProtocolUtils.generateUnsignedPayloadSigV4Header(context, operation);
-    }
-
-    @Override
-    protected void writeDefaultErrorHeaders(GenerationContext context, StructureShape error) {
-        super.writeDefaultErrorHeaders(context, error);
-        context.getWriter().write("'x-amzn-errortype': $S,", error.getId().getName());
+    protected void writeDefaultHeaders(GenerationContext context, Shape operationOrError, boolean isInput) {
+        super.writeDefaultHeaders(context, operationOrError, isInput);
+        if (operationOrError.isOperationShape()) {
+            OperationShape operation = operationOrError.asOperationShape().get();
+            if (isInput) {
+                AwsProtocolUtils.generateUnsignedPayloadSigV4Header(context, operation);
+            }
+        } else if (operationOrError.isStructureShape()) {
```

Should we throw an exception in `else` as an assertion?